x/concentrated-liquidity/lp_test.go

```diff
@@ -622,6 +622,326 @@ func (s *KeeperTestSuite) TestWithdrawPosition() {
 	}
 }
 
+func (s *KeeperTestSuite) TestAddToPosition() {
+	defaultTimeElapsed := time.Hour * 24
+	roundingError := sdk.OneInt()
+
+	// These amounts are set based on the actual amounts passed in as inputs
+	// to create position in the default config case (prior to rounding). We use them as
+	// a reference to test rounding behavior when adding to positions.
+	amount0PerfectRatio := sdk.NewInt(998977)
+	amount1PerfectRatio := sdk.NewInt(5000000000)
+
+	tests := map[string]struct {
+		setupConfig *lpTest
+		// when this is set, it overwrites the setupConfig
+		// and gives the overwritten configuration to
+		// the system under test.
+		sutConfigOverwrite      *lpTest
+		createPositionOverwrite bool
+		timeElapsed             time.Duration
+		createLockLocked        bool
+		createLockUnlocking     bool
+		createLockUnlocked      bool
+		lastPositionInPool      bool
+
+		amount0ToAdd sdk.Int
+		amount1ToAdd sdk.Int
+	}{
+		"sanity check: add zero liquidity": {
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Sub(roundingError),
+				// Since we round on the other the asset when we withdraw, asset0 turns into the bottleneck and
+				// thus we cannot use the full amount of asset1. We calculate the output here (rounding up):
+				// https://www.wolframalpha.com/input?i=1517881404.211361448703481013+*+%2870.710678118654752440+-+67.416615162732695594%29
+				amount1Expected: sdk.NewInt(4999996906),
+			},
+			timeElapsed:  defaultTimeElapsed,
+			amount0ToAdd: sdk.ZeroInt(),
+			amount1ToAdd: sdk.ZeroInt(),
+		},
+		"sanity check: add zero liquidity with rounding error added back": {
+			// Note: this test case is intended to demonstrate that the loss when
+			// adding zero liquidity is due to the rounding error turning the truncated
+			// asset into the bottleneck.
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio,
+				amount1Expected: amount1PerfectRatio,
+			},
+			timeElapsed:  defaultTimeElapsed,
+			amount0ToAdd: roundingError,
+			amount1ToAdd: sdk.ZeroInt(),
+		},
+		"add base amount to existing liquidity with perfect ratio": {
+			// setup parameters for creating a pool and position.
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Add(amount0PerfectRatio).Sub(roundingError),
+				// Since we round on the other the asset when we withdraw, asset0 turns into the bottleneck and
+				// thus we cannot use the full amount of asset1. We calculate the output here (rounding up):
+				// https://www.wolframalpha.com/input?i=3035764327.860030912175533748+*+%2870.710678118654752440+-+67.416615162732695594%29
```

Comment:
So `70.710678118654752440` is the default 5k, how do you get `4545.0`? Could you add a breakdown of that please?

Reply:
Just to clarify, where do you see `4545.0`? The wolfram link simply recalculates the asset1 amount using `asset1 = L * (sqrtPriceUpper - sqrtPriceLower)` (added clarifying comments for this). For context, I pulled the sqrtPrices from the actual positions that were created.